cumulusci/tasks/datadictionary.py

```diff
@@ -339,35 +347,67 @@ def _should_process_object(
 
         return (
-            (sobject_name.endswith("__c") or sobject_name.count("__") == 0)
+            (
+                sobject_name.endswith("__c")
+                or sobject_name.endswith("__mdt")
+                or sobject_name.endswith("__e")
+                or sobject_name.count("__") == 0
+            )
             and sobject_name.startswith(namespace)
-            and element.find("customSettingsType") is None
-            if element
-            else True
+            and (
+                not self._is_protected_object(element)
+                or self.options["include_protected_schema"]
+            )
         )
 
     def _should_process_object_fields(
         self, sobject_name: str, element: Optional[metadata_tree.MetadataElement]
-    ):
+    ) -> bool:
         """Determine if we should track fields from this object in the field dictionary.
         The object itself may or may not be included.
 
-        We will track any field on a custom object or standard entity. We will not track
-        fields on Custom Settings, Platform Events, or Custom Metadata Types."""
-        return (sobject_name.endswith("__c") or sobject_name.count("__") == 0) and (
-            element.find("customSettingsType") is None if element else True
+        We will track any field on a custom object, Platform Event,
+        Custom Setting, Custom Metadata Type, or standard entity.
+        We will not track any object's fields that is protected
+        in this package version."""
+        return (
+            sobject_name.endswith("__c")
+            or sobject_name.endswith("__mdt")
+            or sobject_name.endswith("__e")
+            or sobject_name.count("__") == 0
+        ) and (
+            not self._is_protected_object(element)
+            or self.options["include_protected_schema"]
         )
 
+    def _is_protected_object(
+        self, element: Optional[metadata_tree.MetadataElement]
+    ) -> bool:
+        if not element:
+            return False
+
+        # customSettingsVisibility is used in older API versions.
+        visibility_elem = element.find("customSettingsVisibility")
+        if visibility_elem:
+            return visibility_elem.text != "Public"
+
+        visibility_elem = element.find("visibility")
+        if visibility_elem:
+            return visibility_elem.text != "Public"
+
+        return False
+
     def _process_sfdx_release(self, zip_file: ZipFile, version: PackageVersion):
         """Process an SFDX ZIP file for objects and fields"""
         for f in zip_file.namelist():
             path = PurePosixPath(f)
-            if f.startswith("force-app/main/default/objects"):
+            # Be flexible about processing directories in SFDX context.
+            # This may not be optimal if the repo contains multiple
+            # 2GP package subdirectories.
+            if "objects/" in f and not f.startswith("unpackaged/"):
```

Comment:
Since we know that `sfdx-project.json` is present we have the option of using `packageDirectories` directly, no?

Reply:
Yes. I shied away from it because I'd have to make a call on whether to process only the default package directory, all of them, or provide options to specify, which was more complexity than I wanted to add at the moment. I think it's clearly the right direction to go, though.